lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+//     and throw an error.
+//  4. If its parent is NOT VALIDATED and NOT VALIDATE_FAILED, we recursively call this function on its parent.
+//  5. If after calling this function on its parent, the parent is VALIDATE_FAILED, we add the block to the block index
+//     with status VALIDATE_FAILED and throw an error.
+//  6. If after calling this function on its parent, the parent is VALIDATED, we perform all other validations on the block.
+//
+// The recursive function's invariant is described as follows:
+//   - Base case: If block is VALIDATED or VALIDATE_FAILED, return the BlockNode as-is.
+//   - Recursive case: If the block is not VALIDATED or VALIDATE_FAILED in the blockIndexByHash, we will perform all
+//     validations and add the block to the block index with the appropriate status (VALIDATED OR VALIDATE_FAILED) and
+//     return the new BlockNode.
+//   - Error case: Something goes wrong that doesn't result in the block being marked VALIDATE or VALIDATE_FAILED. In
+//     this case, we will add the block to the block index with status STORED and return the BlockNode.
+func (bc *Blockchain) validateAndIndexBlockPoS(block *MsgDeSoBlock) (*BlockNode, error) {
+	blockHash, err := block.Header.Hash()
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem hashing block %v", block)
+	}
+
+	// Base case - Check if the block is validated or validate failed. If so, we can return early.
+	blockNode, exists := bc.blockIndexByHash[*blockHash]
+	if exists && (blockNode.IsValidateFailed() || blockNode.IsValidated()) {
+		return blockNode, nil
+	}
+
+	// Run the validation for the parent and update the block index with the parent's status. We first
+	// check if the parent has a cached status. If so, we use the cached status. Otherwise, we run
+	// the full validation algorithm on it, then index and an use the result.
+	parentBlockNode, err := bc.validatePreviouslyIndexedBlockPoS(block.Header.PrevBlockHash)
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem validating previously indexed block: ")
+	}
+
+	// Here's where it gets a little tricky. If the parent has a status of ValidateFailed, then we know we store
+	// this block as ValidateFailed. If the parent is not ValidateFailed, we ONLY store the block and move on.
+	// We don't want to store it as ValidateFailed because we don't know if it's actually invalid.
+	if parentBlockNode.IsValidateFailed() {
+		return bc.storeValidateFailedBlockWithWrappedError(block, errors.New("parent block is ValidateFailed"))
+	}
+
+	// If the parent block still has a Stored status, it means that we weren't able to validate it
+	// despite trying. The current block will also be stored as a Stored block.
+	if !parentBlockNode.IsValidated() {
+		return bc.storeValidateFailedBlockWithWrappedError(block, errors.New("parent block is neither Validated nor ValidateFailed"))
```

Comment:
I assume this is because we post-process child blocks after we validate.

Reply:
that's correct - although I *think* I should update this line to simply call `storeBlockInBlockIndex` instead of marking it as validate failed. Although I'm not 100% sure how we'd ever hit this case since `validatePreviouslyIndexedBlockPoS` will either mark the parent block validate or validate failed OR return an error (which gets handled above)

Reply:
resolved in #800